# 결측값(Null) 알아보기

NaN 값에 대하여
- Null과 동일
- Pandas 에서는 Not a Number

In [2]:
import pandas as pd
myfile = 'kpop_idols.csv'
df = pd.read_csv(myfile)

In [4]:
df

Stage Name      Full Name Korean Name K. Stage Name Date of Birth  \
0           A.M  Seong Hyunwoo         성현우           에이엠    1996-12-31   
1           Ace  Jang Wooyoung         장우영           에이스    1992-08-28   
2          Aeji      Kwon Aeji         권애지            애지    1999-10-25   
3          AhIn       Lee Ahin         이아인            아인    1999-09-27   
4          Ahra        Go Ahra         고아라            아라    2001-02-21   
...         ...            ...         ...           ...           ...   
1305        Ziu    Park Heejun         박희준            지우    1997-06-16   
1306         ZN      Bae Jinye         배진예            지엔    1994-06-09   
1307        Zoa     Cho Hyewon         조혜원            조아    2005-05-31   
1308       Zuho       Bae Juho         백주호            주호    1996-07-04   
1309       Zuny      Kim Joomi         김주미            주니    1994-12-08   

             Group      Country Birthplace Other Group Gender  
0        Limitless  South Korea        NaN         NaN      M  
1              VAV  South Korea        NaN         NaN      M  
2          Hashtag  South Korea      Daegu         NaN      F  
3         MOMOLAND  South Korea      Wonju         NaN      F  
4         Favorite  South Korea      Yeosu         NaN      F  
...            ...          ...        ...         ...    ...  
1305           VAV  South Korea        NaN         NaN      M  
1306        LABOUM  South Korea    Bucheon       UNI.T      F  
1307       Weeekly  South Korea        NaN         NaN      F  
1308           SF9  South Korea        NaN         NaN      M  
1309  Ladies' Code  South Korea    Gwangju         NaN      F  

[1310 rows x 10 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1310 entries, 0 to 1309
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Stage Name     1310 non-null   object
 1   Full Name      1304 non-null   object
 2   Korean Name    1304 non-null   object
 3   K. Stage Name  1309 non-null   object
 4   Date of Birth  1310 non-null   object
 5   Group          1219 non-null   object
 6   Country        1310 non-null   object
 7   Birthplace     689 non-null    object
 8   Other Group    122 non-null    object
 9   Gender         1310 non-null   object
dtypes: object(10)
memory usage: 102.5+ KB


df.info()로 대략 알아볼 수 있는데, <br>
위에는 1310 entries가 있으나, Non-Null Count가 그 수와 맞지 않는 만큼 NaN이 있다고 생각할 수 있음.

## 결측 값 다루기
Boolean 인덱싱 에서 True이면 NaN

In [6]:
df.isna()

Stage Name  Full Name  Korean Name  K. Stage Name  Date of Birth  Group  \
0          False      False        False          False          False  False   
1          False      False        False          False          False  False   
2          False      False        False          False          False  False   
3          False      False        False          False          False  False   
4          False      False        False          False          False  False   
...          ...        ...          ...            ...            ...    ...   
1305       False      False        False          False          False  False   
1306       False      False        False          False          False  False   
1307       False      False        False          False          False  False   
1308       False      False        False          False          False  False   
1309       False      False        False          False          False  False   

      Country  Birthplace  Other Group  Gender  
0       False        True         True   False  
1       False        True         True   False  
2       False       False         True   False  
3       False       False         True   False  
4       False       False         True   False  
...       ...         ...          ...     ...  
1305    False        True         True   False  
1306    False       False        False   False  
1307    False        True         True   False  
1308    False        True         True   False  
1309    False       False         True   False  

[1310 rows x 10 columns]

In [7]:
df['Birthplace'].isnull()

0        True
1        True
2       False
3       False
4       False
        ...  
1305     True
1306    False
1307     True
1308     True
1309    False
Name: Birthplace, Length: 1310, dtype: bool

In [8]:
df['Birthplace'][df['Birthplace'].isnull()]

0       NaN
1       NaN
8       NaN
9       NaN
10      NaN
       ... 
1300    NaN
1301    NaN
1305    NaN
1307    NaN
1308    NaN
Name: Birthplace, Length: 621, dtype: object

In [9]:
df['Birthplace'][df['Birthplace'].isna()]

0       NaN
1       NaN
8       NaN
9       NaN
10      NaN
       ... 
1300    NaN
1301    NaN
1305    NaN
1307    NaN
1308    NaN
Name: Birthplace, Length: 621, dtype: object

## NaN이 아닌 값 인덱싱

In [10]:
df['Birthplace'].notnull()

0       False
1       False
2        True
3        True
4        True
        ...  
1305    False
1306     True
1307    False
1308    False
1309     True
Name: Birthplace, Length: 1310, dtype: bool

In [11]:
df['Birthplace'][df['Birthplace'].notnull()]

2         Daegu
3         Wonju
4         Yeosu
5         Seoul
6        Pohang
         ...   
1302      Mokpo
1303      Wuhan
1304      Seoul
1306    Bucheon
1309    Gwangju
Name: Birthplace, Length: 689, dtype: object

## Loc를 이용한 인덱싱

In [12]:
df.loc[df['Birthplace'].notnull()]

Stage Name     Full Name Korean Name K. Stage Name Date of Birth  \
2          Aeji     Kwon Aeji         권애지            애지    1999-10-25   
3          AhIn      Lee Ahin         이아인            아인    1999-09-27   
4          Ahra       Go Ahra         고아라            아라    2001-02-21   
5       Ahyoung   Cho Jayoung         조자영            아영    1991-05-26   
6        Ahyung  Kang Ahyoung         강아형            아형    1996-08-27   
...         ...           ...         ...           ...           ...   
1302       Zelo  Choi Junhong         최준홍            젤로    1996-10-15   
1303    Zhou Mi       Zhou Mi          조미            조미    1986-04-19   
1304       Zico      Woo Jiho         우지호            지코    1992-09-14   
1306         ZN     Bae Jinye         배진예            지엔    1994-06-09   
1309       Zuny     Kim Joomi         김주미            주니    1994-12-08   

               Group      Country Birthplace Other Group Gender  
2            Hashtag  South Korea      Daegu         NaN      F  
3           MOMOLAND  South Korea      Wonju         NaN      F  
4           Favorite  South Korea      Yeosu         NaN      F  
5         Dal Shabet  South Korea      Seoul         NaN      F  
6              P.O.P  South Korea     Pohang         NaN      F  
...              ...          ...        ...         ...    ...  
1302           B.A.P  South Korea      Mokpo         NaN      M  
1303  Super Junior-M        China      Wuhan         NaN      M  
1304         Block B  South Korea      Seoul         NaN      M  
1306          LABOUM  South Korea    Bucheon       UNI.T      F  
1309    Ladies' Code  South Korea    Gwangju         NaN      F  

[689 rows x 10 columns]

In [15]:
df.loc[df['Birthplace'].isnull(), ['Full Name', 'Date of Birth']]

Full Name Date of Birth
0     Seong Hyunwoo    1996-12-31
1     Jang Wooyoung    1992-08-28
8       Kim Heejung    1991-07-13
9        Heo Yoorim    2000-07-21
10         Kim Seri    1996-12-06
...             ...           ...
1300   Kim Moonyong    1991-08-30
1301  Jeong Jaewook    1993-01-27
1305    Park Heejun    1997-06-16
1307     Cho Hyewon    2005-05-31
1308       Bae Juho    1996-07-04

[621 rows x 2 columns]